src/binding.jl

```diff
@@ -242,52 +260,102 @@ function bind_pattern!(
-            error("$(location.file):$(location.line): The type `$bound_type` has " *
-                  "$(length(field_names)) fields but the pattern expects $len fields.")
+        bound_type = bind_type(location, T, input, binder)
+
+        # First try the extractor, then try the struct type.
+        disjuncts = BoundPattern[]
+        extractor_temp = nothing
+
+        # Try the extractor first.
+        # This only works with positional arguments.
+        # TODO support named tuples
+        if match_positionally
+            # Check if there is an extractor method for the pattern type.
+            methods = Base.methods(Match.extract, (Type{bound_type}, Any,))
+            # There is always at least one method (the default), so we know the extractor
+            # method is implemented if there's at least two methods.
+            if length(methods) >= 2
+                conjuncts = BoundPattern[]
+                # call Match.extract(Val(T), input) and match the result against the tuple of subpatterns
+                fetch = BoundFetchExtractorPattern(location, source, input, bound_type, Any)
+                extractor_temp = push_pattern!(conjuncts, binder, fetch)
+                subpattern, assigned = bind_pattern!(location,
+                    Expr(:tuple, subpatterns...), extractor_temp, binder, assigned)
+                push!(conjuncts, subpattern)
+                push!(disjuncts, BoundAndPattern(location, source, conjuncts))
+            end
         end
 
-        for i in 1:len
-            pat = subpatterns[i]
-            if match_positionally
-                field_name = field_names[i]
-                pattern_source = pat
-            else
-                @assert pat.head == :kw
-                field_name = pat.args[1]
-                pattern_source = pat.args[2]
-                if !(field_name in field_names)
-                    error("$(location.file):$(location.line): Type `$bound_type` has " *
-                          "no field `$field_name`.")
-                end
+        # If the extractor failed, try the field-by-field match.
+        type_conjuncts = BoundPattern[]
+
+        if isnothing(extractor_temp)
+            # Ensure that the extractor actually failed.
+            # Otherwise, we could have an ambiguity between the type pattern and the extractor pattern.
+            if !isnothing(extractor_temp)
+                pattern0 = BoundTypeTestPattern(location, :( Base.Nothing ), extractor_temp, Nothing)
+                push!(type_conjuncts, pattern0)
             end
 
-            field_type = nothing
-            if field_name == match_fieldnames(Symbol)[1]
-                # special case Symbol's hypothetical name field.
-                field_type = String
+            field_names::Tuple = match_fieldnames(bound_type)
+
+            if match_positionally && len != length(field_names)
+                # If the extractor is defined, silently fail if the field-by-field match fails.
```

Comment:
These are strange semantics. Why wouldn't you use the extractor if the type has the wrong arity?

Reply:
Or, if you're not going to use the extractor, give the diagnostic.

Reply:
Or, perhaps delete this line comment that is confusing me.